### PR TITLE
httprule: Add headers to ClientConn

### PR DESCRIPTION
**Pull request by camh-**
Add a constructor for ClientConn that takes Options, and add an Option
to set a HTTP Header on each HTTP request it makes. This allows a user
of the ClientConn to set authorization headers and such where those
headers are not specified in the `.proto` file but need to be set more
dynamically.

Also add an Option for overriding the default HTTPClient used by the
ClientConn.

### CODE REVIEW

File: httprule/client.go [hunk 1]
```diff
@@ -43,6 +95,13 @@ func (c *ClientConn) Invoke(ctx context.Context, method string, args, reply inte
 	if err != nil {
 		return err
 	}
+	if c.header != nil {
```
Comment (juliaogris) [inline httprule/client.go:98]: I think this if is not needed: https://play.golang.com/p/akjM13n9f20

File: httprule/client.go [hunk 2]
```diff
@@ -43,6 +95,13 @@ func (c *ClientConn) Invoke(ctx context.Context, method string, args, reply inte
...
+		for key, vals := range c.header {
+			for _, v := range vals {
+				req.Header.Add(key, v)
```
Comment (juliaogris) [inline httprule/client.go:101]: have you considered adding this to the client.Transport instead so I would also be captured for the streaming (iff we ever implement that)?

File: httprule/client_test.go [hunk 3]
```diff
@@ -117,13 +134,20 @@ func TestErrorStatus(t *testing.T) {
 	require.Equal(t, codes.Unknown, status.Code(err))
 }
 
-func newTestServer(body string, statusCode int) *httptest.Server {
-	h := http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+type testServer struct {
+	*httptest.Server
+	request *http.Request
```
Comment (juliaogris) [inline httprule/client_test.go:139]: This took me a moment to work out - maybe a brief comment like:
"save last request to test server for assertions, e.g. if Header was set correctly "
you will find way better words I know.

**Review by juliaogris — APPROVED** [review]
Looks beautiful to me. So nice to review your code.

File: httprule/client.go (hunk 1 above)
Comment (camh-) [reply httprule/client.go:98]: you are right this is not needed. I remember saying this a couple of weeks ago (you can read from a nil map). Will fix.

File: httprule/client.go (hunk 2 above)
Comment (camh-) [reply httprule/client.go:101]: Yes. I did consider that. Then I read this comment on the `RoundTripper` interface:

> RoundTrip should not attempt to handle higher-level protocol details such as redirects, authentication, or cookies...

That thing we did in vcards apparently was not the right way to do this. This idea of client "middleware" wrapping the roundtripper goes against the docs.

Comment (camh-) [reply httprule/client.go:101]: I guess adding headers may not fall into those categories. At the time I was thinking of only adding the auth header via the RoundTripper if i got an auth required response.

File: httprule/client_test.go (hunk 3 above)
Comment (camh-) [reply httprule/client_test.go:139]: Will do

**Pull request merged**